api/src/main/java/org/apache/iceberg/types/ReassignIds.java

```diff
@@ -49,39 +54,53 @@ public Type struct(Types.StructType struct, Iterable<Type> fieldTypes) {
     Preconditions.checkArgument(sourceType.isStructType(), "Not a struct: %s", sourceType);
 
     Types.StructType sourceStruct = sourceType.asStructType();
+
     List<Types.NestedField> fields = struct.fields();
     int length = fields.size();
 
     List<Type> types = Lists.newArrayList(fieldTypes);
     List<Types.NestedField> newFields = Lists.newArrayListWithExpectedSize(length);
     for (int i = 0; i < length; i += 1) {
       Types.NestedField field = fields.get(i);
-      int sourceFieldId = sourceStruct.field(field.name()).fieldId();
+      Types.NestedField sourceField = sourceStruct.field(field.name());
```

Comment:
Is it possible that `sourceStruct == null`? e.g. if this is a field within a new struct being added?

Reply:
Ok... seems no because `sourceType = field.type();` is being set at line 92